### PR TITLE
Add initial cpu userbenchmark for torchbench

### PR DESCRIPTION
**Pull request by chuanqi129**
Add initial cpu userbenchmark for torchbench

Works for Roadmap #1293 for cpu userbenchmark extend with below functions.

- [x] Add core binding option, support multi-instances test.
- [x] Add gomp/iomp option.
- [x] Add memory allocator option.
- [x] Support all enabled cpu features test based on torchbench models, e.g. channels-last / fx_int8 / jit with fusers 
- [x] Support latency and cpu_peak_mem metrics for now, will extend to fps-like report
- [x] Add `README.md`

For example, in below cml, we tested 2 models fx_int8 inference with batch size 8 on CLX socket 0 and 4 instances at the same time.
```shell
$ python run_benchmark.py cpu --model resnet50,alexnet --test eval -b 8 --precision fx_int8 --launcher --launcher-args "--node-id 0 --ninstances 4"
Running benchmark: ['/localdisk/chuanqiw/miniconda3/envs/torchdynamo/bin/python', '-m', 'torch.backends.xeon.run_cpu', '--node-id', '0', '--ninstances', '4', '/localdisk/chuanqiw/PT/benchmark/userbenchmark/cpu/run_config.py', '-m', 'resnet50', '--device', 'cpu', '-b', '8', '-t', 'eval', '-o', PosixPath('/localdisk/chuanqiw/PT/benchmark/.userbenchmark/cpu/cpu-20230420004336')]
2023-04-20 00:43:37,960 - __main__ - INFO - Use JeMalloc memory allocator
2023-04-20 00:43:37,960 - __main__ - INFO - OMP_NUM_THREADS=7
2023-04-20 00:43:37,960 - __main__ - INFO - Using Intel OpenMP
2023-04-20 00:43:37,960 - __main__ - INFO - KMP_AFFINITY=granularity=fine,compact,1,0
2023-04-20 00:43:37,960 - __main__ - INFO - KMP_BLOCKTIME=1
2023-04-20 00:43:37,960 - __main__ - INFO - LD_PRELOAD=/localdisk/chuanqiw/miniconda3/envs/torchdynamo/lib/libiomp5.so:/localdisk/chuanqiw/miniconda3/envs/torchdynamo/lib/libjemalloc.so
2023-04-20 00:43:37,960 - __main__ - INFO - numactl -C 0-6 -m 0 /localdisk/chuanqiw/miniconda3/envs/torchdynamo/bin/python -u /localdisk/chuanqiw/PT/benchmark/userbenchmark/cpu/run_config.py -m resnet50 --device cpu -b 8 -t eval -o /localdisk/chuanqiw/PT/benchmark/.userbenchmark/cpu/cpu-20230420004336
2023-04-20 00:43:37,960 - __main__ - INFO - numactl -C 7-13 -m 0 /localdisk/chuanqiw/miniconda3/envs/torchdynamo/bin/python -u /localdisk/chuanqiw/PT/benchmark/userbenchmark/cpu/run_config.py -m resnet50 --device cpu -b 8 -t eval -o /localdisk/chuanqiw/PT/benchmark/.userbenchmark/cpu/cpu-20230420004336
2023-04-20 00:43:37,960 - __main__ - INFO - numactl -C 14-20 -m 0 /localdisk/chuanqiw/miniconda3/envs/torchdynamo/bin/python -u /localdisk/chuanqiw/PT/benchmark/userbenchmark/cpu/run_config.py -m resnet50 --device cpu -b 8 -t eval -o /localdisk/chuanqiw/PT/benchmark/.userbenchmark/cpu/cpu-20230420004336
2023-04-20 00:43:37,960 - __main__ - INFO - numactl -C 21-27 -m 0 /localdisk/chuanqiw/miniconda3/envs/torchdynamo/bin/python -u /localdisk/chuanqiw/PT/benchmark/userbenchmark/cpu/run_config.py -m resnet50 --device cpu -b 8 -t eval -o /localdisk/chuanqiw/PT/benchmark/.userbenchmark/cpu/cpu-20230420004336
Running TorchBenchModelConfig(name='resnet50', device='cpu', test='eval', batch_size=8, jit=False, extra_args=[], extra_env=None) ...Running TorchBenchModelConfig(name='resnet50', device='cpu', test='eval', batch_size=8, jit=False, extra_args=[], extra_env=None) ...Running TorchBenchModelConfig(name='resnet50', device='cpu', test='eval', batch_size=8, jit=False, extra_args=[], extra_env=None) ...Running TorchBenchModelConfig(name='resnet50', device='cpu', test='eval', batch_size=8, jit=False, extra_args=[], extra_env=None) ... [Done]
 [Done]
 [Done]
 [Done]
Running benchmark: ['/localdisk/chuanqiw/miniconda3/envs/torchdynamo/bin/python', '-m', 'torch.backends.xeon.run_cpu', '--node-id', '0', '--ninstances', '4', '/localdisk/chuanqiw/PT/benchmark/userbenchmark/cpu/run_config.py', '-m', 'alexnet', '--device', 'cpu', '-b', '8', '-t', 'eval', '-o', PosixPath('/localdisk/chuanqiw/PT/benchmark/.userbenchmark/cpu/cpu-20230420004336')]
2023-04-20 00:43:53,444 - __main__ - INFO - Use JeMalloc memory allocator
2023-04-20 00:43:53,444 - __main__ - INFO - OMP_NUM_THREADS=7
2023-04-20 00:43:53,444 - __main__ - INFO - Using Intel OpenMP
2023-04-20 00:43:53,444 - __main__ - INFO - KMP_AFFINITY=granularity=fine,compact,1,0
2023-04-20 00:43:53,444 - __main__ - INFO - KMP_BLOCKTIME=1
2023-04-20 00:43:53,444 - __main__ - INFO - LD_PRELOAD=/localdisk/chuanqiw/miniconda3/envs/torchdynamo/lib/libiomp5.so:/localdisk/chuanqiw/miniconda3/envs/torchdynamo/lib/libjemalloc.so
2023-04-20 00:43:53,445 - __main__ - INFO - numactl -C 0-6 -m 0 /localdisk/chuanqiw/miniconda3/envs/torchdynamo/bin/python -u /localdisk/chuanqiw/PT/benchmark/userbenchmark/cpu/run_config.py -m alexnet --device cpu -b 8 -t eval -o /localdisk/chuanqiw/PT/benchmark/.userbenchmark/cpu/cpu-20230420004336
2023-04-20 00:43:53,445 - __main__ - INFO - numactl -C 7-13 -m 0 /localdisk/chuanqiw/miniconda3/envs/torchdynamo/bin/python -u /localdisk/chuanqiw/PT/benchmark/userbenchmark/cpu/run_config.py -m alexnet --device cpu -b 8 -t eval -o /localdisk/chuanqiw/PT/benchmark/.userbenchmark/cpu/cpu-20230420004336
2023-04-20 00:43:53,445 - __main__ - INFO - numactl -C 14-20 -m 0 /localdisk/chuanqiw/miniconda3/envs/torchdynamo/bin/python -u /localdisk/chuanqiw/PT/benchmark/userbenchmark/cpu/run_config.py -m alexnet --device cpu -b 8 -t eval -o /localdisk/chuanqiw/PT/benchmark/.userbenchmark/cpu/cpu-20230420004336
2023-04-20 00:43:53,445 - __main__ - INFO - numactl -C 21-27 -m 0 /localdisk/chuanqiw/miniconda3/envs/torchdynamo/bin/python -u /localdisk/chuanqiw/PT/benchmark/userbenchmark/cpu/run_config.py -m alexnet --device cpu -b 8 -t eval -o /localdisk/chuanqiw/PT/benchmark/.userbenchmark/cpu/cpu-20230420004336
Running TorchBenchModelConfig(name='alexnet', device='cpu', test='eval', batch_size=8, jit=False, extra_args=[], extra_env=None) ...Running TorchBenchModelConfig(name='alexnet', device='cpu', test='eval', batch_size=8, jit=False, extra_args=[], extra_env=None) ...Running TorchBenchModelConfig(name='alexnet', device='cpu', test='eval', batch_size=8, jit=False, extra_args=[], extra_env=None) ...Running TorchBenchModelConfig(name='alexnet', device='cpu', test='eval', batch_size=8, jit=False, extra_args=[], extra_env=None) ... [Done]
 [Done]
 [Done]
 [Done]
```
We can find the test results in `.userbenchmark/cpu/cpu-20230420004336`, `cpu` userbenchmark will create a subfolder for each test, and aggregate all test results into `metrics-20230420004336.json`. For each sub-folder, it contains instances logs named with instance PID for that model test.  
```shell
$ ls .userbenchmark/cpu/cpu-20230420004336
eval_alexnet_eager/  eval_resnet50_eager/  
$ ls .userbenchmark/cpu/cpu-20230420004336/eval_alexnet_eager/
metrics-3347653.json  metrics-3347654.json  metrics-3347655.json  metrics-3347656.json
$ cat .userbenchmark/cpu/metrics-20230420004336.json 
{
    "name": "cpu",
    "environ": {
        "pytorch_git_version": "de1114554c38322273c066c091d455519d45472d"
    },
    "metrics": {
        "alexnet-eval-eager_latency": 58.309660750000006,
        "alexnet-eval-eager_cmem": 0.416259765625,
        "resnet50-eval-eager_latency": 335.04970325,
        "resnet50-eval-eager_cmem": 0.90673828125
    }
}
```

### CODE REVIEW

**Comment by xuzhao9** [comment]
Can you please rebase on the latest main branch to fix the CI?

**Comment by chuanqi129** [comment]
> Can you please rebase on the latest main branch to fix the CI?

Rebased. Please help to review it when you have time~

File: userbenchmark/cpu/run_config.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+def run_config(config: TorchBenchModelConfig, dryrun: bool=False) -> Optional[TorchBenchModelMetrics]:
+    """This function only handles NotImplementedError, all other errors will fail."""
+    metrics = get_metrics(config)
+    print(f"Running {config} ...", end='')
+    if dryrun:
+        return None
+    # We do not allow RuntimeError in this test
+    try:
+        # load the model instance within the same process
```
Comment (xuzhao9) [inline userbenchmark/cpu/run_config.py:63]: To clarify: `load_model_isolated()` loads model in separate subprocesses.

File: userbenchmark/cpu/__init__.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    parser.add_argument("--test", "-t", default="eval", help="Tests to run, splited by comma.")
+    parser.add_argument("--model", "-m", default=None, help="Only run the specifice models, splited by comma.")
+    parser.add_argument("--batch-size", "-b", default=None, help="Run the specifice batch size.")
+    parser.add_argument("--jit", action="store_true", help="Convert the models to jit mode.")
+    parser.add_argument("--config", "-c", default=None, help="YAML config to specify tests to run.")
+    parser.add_argument("--output", "-o", default=None, help="Output dir.")
+    parser.add_argument("--launcher", action="store_true", help="Use torch.backends.xeon.run_cpu to get the peak performance on Intel(R) Xeon(R) Scalable Processors.")
+    parser.add_argument("--launcher-args", default=None, help="Provide the args of torch.backends.xeon.run_cpu. See `python -m torch.backends.xeon.run_cpu --help`")
+    parser.add_argument("--dryrun", action="store_true", help="Dryrun the command.")
+    return parser.parse_known_args()
```
Comment (xuzhao9) [inline userbenchmark/cpu/__init__.py:90]: Should this be `parser.parse_known_args(args)`?

File: userbenchmark/cpu/__init__.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+def run(args: List[str]):
+    args, extra_args = parse_args(args)
+    extra_args.remove(BM_NAME)
```
Comment (xuzhao9) [inline userbenchmark/cpu/__init__.py:94]: This can be removed, if we only parse `args`.

**Review by xuzhao9 — COMMENTED** [review]
If the userbenchmark has only one output json metrics file, we expect it to be named as `metrics-YYmmddHHMMSS.json`, where YYmmddHHMMSS is the time the file is written. It will be much easier for us to implement regression detection and regression bisection on this.

File: userbenchmark/cpu/__init__.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
+"""
```
[the remark this answers is not in the document]
Comment (chuanqi129) [reply userbenchmark/cpu/__init__.py:1]: Good point. Will add it

File: userbenchmark/cpu/run_config.py (hunk 1 above)
Comment (chuanqi129) [reply userbenchmark/cpu/run_config.py:63]: Is there any potential issue in here? I have double checked the numactl setting should be apply to subprocess also, am I right?

File: userbenchmark/cpu/__init__.py (hunk 3 above)
Comment (chuanqi129) [reply userbenchmark/cpu/__init__.py:94]: Correct, thanks to point it out, will update it

File: userbenchmark/cpu/__init__.py (hunk 2 above)
Comment (chuanqi129) [reply userbenchmark/cpu/__init__.py]: Modified

**Comment by chuanqi129** [comment]
> If the userbenchmark has only one output json metrics file, we expect it to be named as `metrics-YYmmddHHMMSS.json`, where YYmmddHHMMSS is the time the file is written. It will be much easier for us to implement regression detection and regression bisection on this.

Got your point, modified the summary metrics file name from `cpu_res.json` to `metrics-YYmmddHHMMSS.json`, but the YYmmddHHMMSS is the time to start the tests. It can help to align the subdir folder name with metrics json name on the suffix. How do you think?

File: userbenchmark/cpu/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (chuanqi129) [reply userbenchmark/cpu/__init__.py:1]: [`README.md`](https://github.com/chuanqi129/benchmark/blob/chuanqiw/cpu_userbenchmark/userbenchmark/cpu/README.md) has been added, please help to review again

**Comment by xuzhao9** [comment]
> > Thanks for raise the potential issue early, do you think that enlarge warmup iters and test iters by parameters is a good solution?


I am not sure about that yet, we need to actually run the benchmark to understand the noise level. But we should have feature of measuring of noise levels in the benchmark.

> 
> > From features coverage perspective, if we have CPX / SPR instance will be great, because those platform support FP32/BF16/INT8 datatype. But at the beginning, I think fp32 / int8 basic test is enough, that means ICX / CLX can meet the requirements. How do you think?

Sounds good to me. I can help setup the permission to GitHub Actions if you don't have it yet.

> 
> > > We can try both on-demand AWS instances and our fine-tuned AWS metal instance, and see how different the variations are.
> > 
> > 
> > Sounds good. How could we enable test on those instance?

The AWS metal instance is AWS g4dn.metal, which is Intel(R) Xeon(R) Platinum 8259CL CPU, does it support fp32/int8, or CPX/SPR? If not, we need to ask @malfet what AWS CPU runners are available.

File: userbenchmark/cpu/run_config.py (hunk 1 above)
Comment (xuzhao9) [reply userbenchmark/cpu/run_config.py:63]: Running multiple models within the same process is known to be problematic on GPU (for example, it won't release GPU memory after running.), but I am not sure if it is the same case on CPU.
Here I am just saying if you decide to run in subprocesses, the comment text needs to be updated to avoid confusion.

File: userbenchmark/cpu/run_config.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+import argparse
+import os
+import sys
+import numpy
+
+from typing import List, Tuple, Dict, Optional
+from pathlib import Path
+
+REPO_DIR = str(Path(__file__).parent.parent.parent.resolve())
+sys.path.append(REPO_DIR)
```
Comment (malfet) [inline userbenchmark/cpu/run_config.py:14]: If you are using `with add_path` later, why set it globally?
```suggestion
```

File: userbenchmark/cpu/cpu_utils.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+Run PyTorch cpu benchmarking.
+"""
+import json
+import os
+import re
+import time
+from datetime import datetime
+from pathlib import Path
+
+REPO_PATH = Path(os.path.abspath(__file__)).parent.parent.parent
```
Comment (malfet) [inline userbenchmark/cpu/cpu_utils.py:11]:
```suggestion
REPO_PATH = Path(__file__).absolute().parent.parent.parent
```

File: userbenchmark/cpu/cpu_utils.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+USERBENCHMARK_OUTPUT_PREFIX = ".userbenchmark"
+
+def get_output_dir(bm_name, test_date=None):
+    current_dir = Path(os.path.dirname(os.path.abspath(__file__)))
```
Comment (malfet) [inline userbenchmark/cpu/cpu_utils.py:15]:
```suggestion
    current_dir = Path(__file__).parent.absolute()
```

File: userbenchmark/cpu/run_config.py (hunk 1 above)
Comment (chuanqi129) [reply userbenchmark/cpu/run_config.py:63]: Got you. Thanks, will update the comment text firstly

File: userbenchmark/cpu/run_config.py (hunk 5 above)
Comment (chuanqi129) [reply userbenchmark/cpu/run_config.py]: Due to we need use the `add_path` class in `utils`. But I have copy the `add_path` into `cpu_utils.py` also, we can remove the this line.

**Comment by xuzhao9** [comment]
Similar to https://github.com/pytorch/benchmark/blob/main/.github/workflows/userbenchmark-regression-detector.yml, we can add a GitHub actions workflow which will run the benchmark on the PyTorch nightly build. But we need to determine the runner label, which corresponds to an AWS instance type.

**Comment by chuanqi129** [comment]
Thanks @xuzhao9 for your great support. And Sorry for the late reply de to out of office. Some reply as below

>But we should have feature of measuring of noise levels in the benchmark.

Agree with you, I will try to implement such feature in another PR.

>For on-demand AWS instances, can you check any of the AWS instances in https://github.com/pytorch/test-infra/blob/main/.github/scale-config.yml can be used? It is the preferred approach.

I have double checked the instance types in Pytorch ci node pool, `c5` instance based on 2nd Generation Intel® Xeon® Scalable Processors (CLX), which can support fp32 and int8, but no bf16 datatype support. We can try to use `linux.24xlarge` instance for initial test.

> If not (for example, we can't reach a reasonable low noise level), we can use the AWS metal instance, which is AWS g4dn.metal, with Intel(R) Xeon(R) Platinum 8259CL CPU. Does it support fp32/int8?

The 8259CL also belongs to 2nd Generation Intel® Xeon® Scalable Processors (CLX). So it also support fp32/int8. We can use `linux.24xlarge` instance firstly, if it has large noise, we can try this metal one later.
